lib/services/vote_animation_service.dart

```diff
@@ -0,0 +1,135 @@
+  unfocused,
+  voteFocused,
+  voteUnfocused,
+  clearFocused,
+  clearUnfocused,
+}
+
+typedef VoteItemFn = Future Function(Vote vote);
+typedef VoteStateChangedHandler = void Function(VoteAnimation voteAnimation);
+
+class VoteAnimationService {
+  final VoteItemFn voteItemHandler;
+  final Vote initialVote;
+  Vote lastVote;
+
+  VoteAnimationService({this.voteItemHandler, this.initialVote})
+      : lastVote = initialVote;
+
+  final Map<VoteButtonType, ValueNotifier<VoteAnimation>> buttonStates =
+      new Map();
+
+  Color getInitialColor(VoteButtonType type) {
+    VoteAnimation initialState = _reduceState(
+      vote: initialVote,
+      button: type.toVote(),
+      whenFocused: VoteAnimation.focused,
+      whenVoted: VoteAnimation.voted,
+      whenElse: VoteAnimation.unfocused,
+    );
+    return (initialState == VoteAnimation.focused ||
+            initialState == VoteAnimation.clearFocused)
+        ? focusedColor
+        : (initialState == VoteAnimation.voted ||
+                initialState == VoteAnimation.voteFocused ||
+                initialState == VoteAnimation.voteUnfocused)
+            ? (type == VoteButtonType.down) ? downVotedColor : votedColor
+            : unfocusedColor;
+  }
+
+  Future voteItem(Vote vote) async {
+    if (vote == lastVote) {
+      vote = vote == Vote.favorite ? Vote.up : Vote.none;
+    } else if (vote == Vote.up && lastVote == Vote.favorite) vote = Vote.none;
+    _setStates(vote);
+    lastVote = vote;
+  }
+
+  void addButtonStateListener(
+      VoteButtonType type, VoteStateChangedHandler onStateChange) {
+    if (buttonStates[type] == null)
+      buttonStates[type] = ValueNotifier<VoteAnimation>(VoteAnimation.initial)
+        ..addListener(() => onStateChange(buttonStates[type].value));
+    _setState(type, lastVote == null ? initialVote : lastVote);
+  }
+
+  void _setStates(Vote vote) =>
+      buttonStates.keys.forEach((type) => _setState(type, vote));
+
+  void _setState(VoteButtonType type, Vote vote) => buttonStates[type].value =
+      _getState(type.toVote(), buttonStates[type].value, vote);
+
+  VoteAnimation _getState(Vote button, VoteAnimation currentState, Vote vote) {
+    if ([
+      VoteAnimation.voted,
+      VoteAnimation.voteFocused,
+      VoteAnimation.voteUnfocused,
+    ].contains(currentState))
+      return _reduceState(
+        vote: vote,
+        button: button,
+        whenFocused: VoteAnimation.clearFocused,
+        whenVoted: VoteAnimation.voted,
+        whenElse: VoteAnimation.clearUnfocused,
+      );
+
+    if ([
+      VoteAnimation.focused,
+      VoteAnimation.clearFocused,
+    ].contains(currentState))
+      return _reduceState(
+        vote: vote,
+        button: button,
+        whenFocused: VoteAnimation.focused,
+        whenVoted: VoteAnimation.voteFocused,
+        whenElse: VoteAnimation.unfocused,
+      );
+
+    if ([
+      VoteAnimation.unfocused,
+      VoteAnimation.clearUnfocused,
+    ].contains(currentState))
```

Comment:
If too large use braces to make it more readable

Reply:
fixed and refactored with 44437ea